resourceIndexer/core-url-remapping.js

```diff
@@ -0,0 +1,49 @@
+import { readFileSync } from "fs";
+import { convertManifestJsonToAvailableDatasetList } from "../src/endpoints/charon/setAvailableDatasets.js";
+
+let resolvedPath;
+
+/**
+ * Remaps an incomplete URL core dataset path to the path it will be redirected
+ * to by the server via the "nextstrain manifest". If the provided URL would not
+ * be redirected we return the provided path unchanged.
+ * @param {string} path nextstrain.org URL path
+ * @returns {string}
+ */
+export const remapCoreUrl = (path) => {
+  if (!resolvedPath) {
+    /* map only computed the first time this function is run */
+    const serverManifest = JSON.parse(readFileSync("./data/manifest_core.json"));
+    resolvedPath = resolveAll(
+      convertManifestJsonToAvailableDatasetList(serverManifest).defaults
+    )
+  }
+  return resolvedPath[path] || path;
+}
+
+
+/**
+ * The server contains a map of partial paths and their next (default) path,
+ * e.g. flu → seasonal
+ *      flu/seasonal → h3n2
+ *      etc
+ * and uses an iterative approach to resolve the final path. Here we want to
+ * resolve the entire path for every partial path, for instance
+ *      flu → flu/seasonal/h3n2/h2/2y
+ *      flu/seasonal → flu/seasonal/h3n2/h2/2y
+ * @param {Object} defaults 
+ * @returns {Object}
+ */
+function resolveAll(defaults) {
+  const urlMap = {};
+  const defaultKeys = new Set(Object.keys(defaults));
+  Object.entries(defaults).forEach(([partialPath, nextPathPart]) => {
```

Comment:
_non-blocking_

 `.forEach()` with side-effects is always odd to me...

```suggestion
  for (const [partialPath, nextPathPart] of Object.entries(defaults)) {
```

would be what I think is the more typical way to write this imperative code.

Reply:
Thanks. `.forEach()` without side-effects would seem odd to me - it would be doing nothing! I'll keep this advice in mind for future work.

Reply:
Right. It can sometimes be convenient in an interactive context for inspection, but outside a REPL I find there's generally ~no upside to the indirection of the `.forEach()` methods (only downside) vs. the typical imperative `for` loops.